### DISCUSSION TITLE
Hitting a wall getting animation to work

### DISCUSSION THREAD

**Discussion by raarts** [xyflow/xyflow#1992] [Q&A] [answered]
Hi,

I am using dagre for layouting, and I want to animate it. After layouting I end up with two arrays of nodes and edges (the original and the ones generated by dagre), but I can't seem to animate between them. I'm using custom nodes.

There are many animation libraries of course (already using a tiny bit of react-spring in my app, though not in the example below), but I'm also looking at plain css because I found that suggestion in one of the issues.

So here's my question: can anyone provide a general approach on how to tackle this? Are there things I need to watch out for, with respect to react-flow?  

Example: I defined custom nodes, and in that node I import css that looks like this:

```css
.knowledge {
    transition: all 0.5s ease-out;
    border-style: solid;
    border-radius: 3px;
    padding: 4px;
    display: flex;
    flex-direction: column;
    align-items: center;
    min-width: 120px;
}
```

My component looks like: 

```JSX
    <div className="knowledge">
      <Handle type="target" position={targetPosition} id="a" />
      <div>
        <div>{data.summary}</div>
      </div>
      <Handle type="source" position={sourcePosition} id="b" />
    </div>
```

But the transition isn't working (all nodes immediately snap to the new position). My layout code (which in itself works flawlessly) looks like:

```Typescript
  const { nodes: layoutedNodes, edges: layoutedEdges } = getDagreLayoutedElements(nodes, edges, 'LR');
  setNodes(layoutedNodes);
  setEdges(layoutedEdges);
```

(I also tried `setNodes([...layoutedNodes])` to no avail).

Here's my Dagre layout function:

```Typescript
export const getDagreLayoutedElements = (nodes: Node<any>[], edges: Edge<any>[], direction = 'LR') => {
  const isHorizontal = direction === 'LR';
  let nodeWidth = 153;
  let nodeHeight = 44;

  const dagreGraph = new dagre.graphlib.Graph();
  dagreGraph.setDefaultEdgeLabel(() => ({}));
  dagreGraph.setGraph({ rankdir: direction });

  nodes.forEach((node) => {
    dagreGraph.setNode(node.id, { width: node.width || nodeWidth, height: node.height || nodeHeight });
    node.targetPosition = isHorizontal ? Position.Left : Position.Top;
    node.sourcePosition = isHorizontal ? Position.Right : Position.Bottom;
  });

  edges.forEach((edge) => {
    dagreGraph.setEdge(edge.source, edge.target);
  });

  dagre.layout(dagreGraph);

  nodes.forEach((node) => {
    const nodeWithPosition = dagreGraph.node(node.id);
    // We are shifting the dagre node position (anchor=center center) to the top left
    // so it matches the React Flow node anchor point (top left).
    console.log('dagre node:', nodeWithPosition);
    console.log('our node:', node);
    node.position = {
      x: nodeWithPosition.x - (node.width || nodeWidth) / 2,
      y: nodeWithPosition.y - (node.height || nodeHeight) / 2,
    };

    return node;
  });

  return { nodes, edges };
};
```

I'm hitting up a wall here, and the fact I have no experience animating using css certainly doesn't help. 
Thanks for any pointers/insights!

**Comment by moklick** [comment, marked as answer]
The node is wrapped with a div that gets the position. You need to add `transition: all 0.5s ease-out` to `.react-flow__node` (or `.react-flow__node-YOUR_CUSTOM_NODE_TYPE`). The issue with this approach is, that the edges are not animated.

**Reply by raarts** [reply]
Thanks, it works now, and I see what you mean. But at least I now know where to act on, I'll try animating the outer div using some JS library, maybe this will improve the edge-animation. Is there a way to get a ref to the outer div?

**Reply by raarts** [reply]
nvm, found it.
